File: be/internal/websocket/events.go
package websocket

// Client → server (JSON text frames unless noted).
const (
	EventSessionStart    = "session_start"
	EventStartListening  = "start_listening"
	EventStopSpeaking    = "stop_speaking"
	EventSessionEnd   = "session_end"
	EventPing         = "ping"
)

// Server → client.
const (
	EventReady             = "ready"
	EventListening         = "listening"
	EventPartialTranscript = "partial_transcript"
	EventFinalTranscript   = "final_transcript"
	EventTypingStart       = "typing_start"
	EventTypingEnd         = "typing_end"
	EventAIResponse        = "ai_response"
	EventSubtitle          = "subtitle"
	EventTranslationDelta  = "translation_delta"
	EventAIAudioChunk       = "ai_audio_chunk"
	EventAIAudioSegmentEnd  = "ai_audio_segment_end"
	EventAIAudioEnd         = "ai_audio_end"
	EventSessionEnded      = "session_ended"
	EventPong              = "pong"
	EventError             = "error"
	EventQuotaExceeded     = "quota_exceeded"
	EventQuotaWarning      = "quota_warning"
)

// ClientMessage control payloads.
type ClientMessage struct {
	Type        string `json:"type"`
	Text        string `json:"text,omitempty"` // final transcript from client STT
	Translation string `json:"translation,omitempty"`
	UserID    int    `json:"user_id,omitempty"`
	SessionID string `json:"session_id,omitempty"`
}

// HistoryMessage is a prior turn sent on connect.
type HistoryMessage struct {
	ID          string `json:"id,omitempty"`
	Role        string `json:"role"`
	Content     string `json:"content"`
	Translation string `json:"translation,omitempty"`
}

// ServerMessage outbound events.
type ServerMessage struct {
	Type       string           `json:"type"`
	Text        string           `json:"text,omitempty"`
	Translation string           `json:"translation,omitempty"`
	SttContext  string           `json:"stt_context,omitempty"`
	Messages   []HistoryMessage `json:"messages,omitempty"`
	Message    string           `json:"message,omitempty"`
	Code       string           `json:"code,omitempty"`
	UserID     int    `json:"user_id,omitempty"`
	SessionID  string `json:"session_id,omitempty"`
	Delta      string `json:"delta,omitempty"`
	FullText   string `json:"full_text,omitempty"`
	Audio      string `json:"audio,omitempty"` // base64 chunk
	Format     string `json:"format,omitempty"`
	Index      int    `json:"index,omitempty"`
	Finished   bool   `json:"finished,omitempty"`
}

File: fe/lib/ws/events.ts
export const ClientEvents = {
  StartListening: "start_listening",
  StopSpeaking: "stop_speaking",
  SessionEnd: "session_end",
  Ping: "ping",
} as const;

export const ServerEvents = {
  Ready: "ready",
  Listening: "listening",
  PartialTranscript: "partial_transcript",
  FinalTranscript: "final_transcript",
  TypingStart: "typing_start",
  TypingEnd: "typing_end",
  AIResponse: "ai_response",
  Subtitle: "subtitle",
  TranslationDelta: "translation_delta",
  AIAudioChunk: "ai_audio_chunk",
  AIAudioSegmentEnd: "ai_audio_segment_end",
  AIAudioEnd: "ai_audio_end",
  SessionEnded: "session_ended",
  Pong: "pong",
  Error: "error",
  QuotaExceeded: "quota_exceeded",
  QuotaWarning: "quota_warning",
} as const;

export type HistoryMessage = {
  id?: string;
  role: "user" | "assistant";
  content: string;
  translation?: string;
};

export type ServerMessage = {
  type: string;
  text?: string;
  translation?: string;
  stt_context?: string;
  messages?: HistoryMessage[];
  history_has_more?: boolean;
  message?: string;
  user_id?: number;
  session_id?: string;
  delta?: string;
  full_text?: string;
  audio?: string;
  format?: string;
  index?: number;
  finished?: boolean;
  code?: string;
};

export type ChatMessage = {
  id: string;
  role: "user" | "assistant";
  content: string;
  translationVi?: string;
  streaming?: boolean;
  /** Chat mode: play one-shot bubble pop when message appears */
  justArrived?: boolean;
};

export type ConnectionStatus =
  | "disconnected"
  | "connecting"
  | "ready"
  | "listening"
  | "thinking"
  | "speaking"
  | "error";
